sys/kern/build.rs

```diff
@@ -45,3 +54,110 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let mut file = File::create(out.join("kconfig.rs")).unwrap();
+
+    writeln!(file, "// See build.rs for details")?;
+
+    writeln!(
+        file,
+        "const HUBRIS_FAULT_NOTIFICATION: u32 = {};",
+        kconfig.supervisor_notification
+    )?;
+    writeln!(
+        file,
+        "const HUBRIS_TASK_COUNT: usize = {};",
+        kconfig.tasks.len()
+    )?;
+
+    writeln!(
+        file,
+        "static HUBRIS_TASK_DESCS: [abi::TaskDesc; {}] = [",
+        kconfig.tasks.len()
+    )?;
+    for task in &kconfig.tasks {
+        writeln!(file, "    abi::TaskDesc {{")?;
+        writeln!(file, "        regions: [")?;
+        for region in &task.regions {
+            writeln!(file, "            {},", region)?;
+        }
+        writeln!(file, "        ],")?;
+        writeln!(file, "        entry_point: 0x{:08x},", task.entry_point)?;
+        writeln!(file, "        initial_stack: 0x{:08x},", task.initial_stack)?;
+        writeln!(file, "        priority: {},", task.priority)?;
+        writeln!(
+            file,
+            "        flags: unsafe {{ \
+            abi::TaskFlags::from_bits_unchecked({}) }},",
+            task.flags.bits()
+        )?;
+        writeln!(file, "    }},")?;
+    }
+    writeln!(file, "];")?;
+
+    writeln!(
+        file,
+        "static mut HUBRIS_TASK_TABLE_SPACE: \
+        core::mem::MaybeUninit<[crate::task::Task; {}]> = \
+        core::mem::MaybeUninit::uninit();",
+        kconfig.tasks.len()
+    )?;
+
+    writeln!(
+        file,
+        "static mut HUBRIS_REGION_TABLE_SPACE: \
+        core::mem::MaybeUninit<[\
+            [&'static abi::RegionDesc; crate::app::REGIONS_PER_TASK]; \
+            HUBRIS_TASK_COUNT]> = core::mem::MaybeUninit::uninit();"
+    )?;
+
+    writeln!(
+        file,
+        "static HUBRIS_REGION_DESCS: [abi::RegionDesc; {}] = [",
+        kconfig.regions.len()
+    )?;
+    for region in &kconfig.regions {
+        writeln!(file, "    abi::RegionDesc {{")?;
+        writeln!(file, "        base: 0x{:08x},", region.base)?;
+        writeln!(file, "        size: 0x{:08x},", region.size)?;
+        writeln!(
+            file,
+            "        attributes: unsafe {{ \
+            abi::RegionAttributes::from_bits_unchecked({}) }},",
+            region.attributes.bits()
+        )?;
+        writeln!(file, "        reserved_zero: 0,")?;
+        writeln!(file, "    }},")?;
+    }
+    writeln!(file, "];")?;
+
+    writeln!(
+        file,
+        "static HUBRIS_INTERRUPTS: [abi::Interrupt; {}] = [",
+        kconfig.irqs.len()
+    )?;
+    for irq in &kconfig.irqs {
+        writeln!(file, "    abi::Interrupt {{")?;
+        writeln!(file, "        irq: {},", irq.irq)?;
+        writeln!(file, "        task: {},", irq.task)?;
+        writeln!(file, "        notification: 0b{:b},", irq.notification)?;
+        writeln!(file, "    }},")?;
+    }
+    writeln!(file, "];")?;
+
+    Ok(())
+}
+
+#[derive(Deserialize)]
+struct KernelConfig {
```

Comment:
Is there an easy way to share this with the `xtask/src/dist.rs` definition?

Reply:
Nooooot right now. :-( I would like to improve this in a followup, we have a number of config data structures that could use to be factored out into a crate.